### PR TITLE
stabilize ec2_vpc_subnet module - Fixes #36083

### LINKED ISSUE #36083: Unstable integration test ec2_vpc_subnet
##### ISSUE TYPE

Bug Report

##### COMPONENT NAME

ec2_vpc_subnet integration test

##### ANSIBLE VERSION

2.5

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

Shippable

##### SUMMARY

Unstable integration test ec2_vpc_subnet:

https://app.shippable.com/github/ansible/ansible/runs/54187/55/tests

##### STEPS TO REPRODUCE

Run tests on Shippable.

##### EXPECTED RESULTS

Tests pass.

##### ACTUAL RESULTS

Tests are unstable or fail.

> test/integration/targets/ec2_vpc_subnet/tasks/main.yml:350 / [localhost] amazon: ec2_vpc_subnet : tidy up subnet aws_secret_key={{ aws_secret_key }}, aws_access_key={{ aws_access_key }}, security_token={{ security_token }}, state=absent, vpc_id={{ vpc_result.vpc.id }}, region={{ ec2_region }}, cidr=10.232.232.128/28

> failure: The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'vpc' The error appears to have been in '/root/ansible/test/integration/targets/ec2_vpc_subnet/tasks/main.yml': line 350, column 7, but may be elsewhere in the file depending on the exact syntax problem. The offending line appears to be: - name: tidy up subnet ^ here

```
{
"failed": true, 
"msg": "The task includes an option with an undefined variable. The error was: 'dict object' has no attribute 'vpc'\n\nThe error appears to have been in '/root/ansible/test/integration/targets/ec2_vpc_subnet/tasks/main.yml': line 350, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n    - name: tidy up subnet\n      ^ here\n"
}
```

> test/integration/targets/ec2_vpc_subnet/tasks/main.yml:14 / [localhost] amazon: ec2_vpc_subnet : create a VPC aws_secret_key={{ aws_secret_key }}, aws_access_key={{ aws_access_key }}, name={{ resource_prefix }}-vpc, security_token={{ security_token }}, tags={u'Name': u'{{ resource_prefix }}-vpc', u'Description': u'Created by ansible-test'}, region={{ ec2_region }}, state=present, cidr_block=10.232.232.128/26

> error: ClientError: An error occurred (InvalidVpcID.NotFound) when calling the DescribeVpcClassicLink operation: The vpc ID 'vpc-1b2bda60' does not exist

```
Traceback (most recent call last):
  File "/tmp/ansible_dPE6PL/ansible_module_ec2_vpc_net.py", line 203, in get_vpc
    classic_link = connection.describe_vpc_classic_link(VpcIds=[vpc_id])['Vpcs'][0].get('ClassicLinkEnabled')
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 317, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 615, in _make_api_call
    raise error_class(parsed_response, operation_name)
ClientError: An error occurred (InvalidVpcID.NotFound) when calling the DescribeVpcClassicLink operation: The vpc ID 'vpc-1b2bda60' does not exist
```

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fixes #36083

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py

##### ANSIBLE VERSION
```
2.6.0
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: Looking for a way to optimize this - watching the AWS console while the tests run, this seems to take much longer than it should.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: I think it'd be best to separately handle `WaiterError` here, so we can provide a message for "timed out waiting for update to [insert what you waited for here]" and another message for "Error XYZ occured while waiting for [whatever]" so that we can see in bug reports whether the wait time is just too short or if there is an error while waiting.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_subnet.py]: Can you please fix the capitalization here also? `s/Subnet/subnet`

**Review by ryansb — CHANGES_REQUESTED** [review]
A couple extra things in addition to the renames we discussed.

**Review by ryansb — APPROVED** [review]

**Comment by willthames** [comment]
@s-hertel good effort with all the new waiters. Will they be incorporated upstream into boto3, do you think?

**Review by willthames — CHANGES_REQUESTED** [review]
Need to fix the issue identified while testing #37925

**Comment by s-hertel** [comment]
Thanks @willthames for catching that! I'll add a test for it as well.

It looks like there have been a few waiters accepted from external contributors in botocore, so I'll open a pull request and see if they want them.

This module also supports check mode but there are no tests for it, so I'm going to add those as well.

**Comment by s-hertel** [comment]
@willthames I've fixed the tagging behavior to be idempotent and fixed the traceback in https://github.com/ansible/ansible/issues/37925#issuecomment-376357452. I also added tests.

The bug report in #37925 is still partially unsolved and is also a bug with any AWS module that has tags and would need a deeper fix than at the module level. `true` is seen as the Python bool True and then stringified to "True" rather than "true", so if the actual tag is lowercase this will report changed: true and then be subsequently idempotent (because it will have added the capitalized tag). I think that would need to be fixed at the parsing level, though it could as easily be considered a misuse of yaml (true really should just be in quotation marks).

**Comment by s-hertel** [comment]
I think this is good-to-go now, but I want to run the tests again to check that the tests are quite stable so please don't merge until I've updated this comment.

Update: I'm getting some sporadic errors still (though fewer than on devel). Don't merge this.

**Comment by willthames** [comment]
@s-hertel I just wanted to make sure that this change doesn't worsen the behaviour when setting a tag with `true` - I agree there's not much we can do, but don't want new stack traces!

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
I've approved this change now, so good to merge when you think it's ready @s-hertel

**Comment by ryansb** [comment]
The only unfixed stability issues we've found are: 

- Mapping public IP addresses is still sometimes unstable for `u'map_public_ip_on_launch'`
- IPv6 settings can still be inconsistent (rarely) for `u'assign_ipv6_address_on_creation'`

These both still happen in this PR, but are much less frequent than on devel.

**Comment by s-hertel** [comment]
bot_status

**Pull request merged**
